// express-base-controller/src/api.controller.ts
import {
  ObjectID,
} from 'bson';
import {
  Response,
  NextFunction,
} from 'express';
import {
  IApiRequest,
} from './types/IApiRequest';
import {
  IApiError,
} from './types/IApiError';
import {
  isString,
} from './helpers/isString';
import {
  IApiDocument,
  IApiModel,
} from './types/IApiModel';

abstract class ApiController<T extends IApiModel<K, R>, K = unknown, R = unknown> {
  protected model: T;

  constructor(model: T) {
    this.model = model;
  }

  // ERROR Responses
  public respondServerError(res: Response, error: any): Response {
    return res.status(500).json({ error });
  }
  public respondNotFound(
    id: string | number | ObjectID,
    res: Response,
    modelName: string,
  ): Response {
    const error: IApiError = {
      id: 'notFound',
      message: `${ modelName } ${ id } does not exist`,
      fields: {},
      errors: [],
    };

    return res.status(404).json({
      error: error,
    });
  }
  public respondInvalidId(res: Response): Response {
    const error: IApiError = {
      id: 'invalidId',
      message: 'Invalid id',
      fields: {},
      errors: [],
    };

    return res.status(400).json({
      error: error,
    });
  }
  public respondModelMissingError(res: Response): Response {
    const error: IApiError = {
      id: 'modelMissing',
      message: 'the model is missing in the request',
      fields: {},
      errors: [],
    };

    return res.status(400).json({
      error: error,
    });
  }
  public respondDeletionError(res: Response, err: IApiError): Response {
    const error: IApiError = {
      id: 'delete',
      message: err.message,
      fields: {},
      errors: [],
    };

    return res.status(400).json({
      error: error,
    });
  }
  public respondValidationError(err: any, res: Response, next: NextFunction): Response | void {
    if (err.name === 'ValidationError') {
      const error: IApiError = {
        id: 'validationError',
        message: err.message,
        fields: err.errors,
        errors: [],
      };

      return res.status(400).json({
        error: error,
      });
    }
    if (err.code === 11000) {
      return res.status(400).json({
        error: {
          id: 'duplicate',
          message: `${ this.model.modelName } already exists`,
        },
      });
    } else {
      return next(err);
    }
  }
  public apiResponse(req: IApiRequest, res: Response, _next: NextFunction): Response {
    const hasMetaError = req.meta && req.meta.error;
    let status = 200;
    if (hasMetaError) {
      status = 500;
    }

    return res.status(status).json({ meta: req.meta, data: req.data });
  }
  public populateMeta(req: IApiRequest, _res: Response, next: NextFunction): void {
    const qTotal = req.modelQuery.total || {};

    this.model.countDocuments(qTotal, (err: any, total: number) => {
      if (err) {
        return next(err);
      } else {
        const offset = req.modelQuery.offset;
        const limit = req.modelQuery.limit;
        req.meta = {
          total: total,
          count: req.data ? req.data.length : 0,
          offset: isString(offset) ? parseInt(offset, 10) : offset,
          limit: isString(limit) ? parseInt(limit, 10) : limit,
        };

        return next();
      }
    });
  }
  protected hasModel(model?: IApiDocument): model is IApiDocument {
    return typeof model !== 'undefined';
  }
}

export default ApiController;
